x-pack/legacy/plugins/siem/server/lib/detection_engine/alerts/utils.test.ts

```diff
@@ -0,0 +1,658 @@
+ * you may not use this file except in compliance with the Elastic License.
+ */
+import { BulkIndexDocumentsParams, SearchParams } from 'elasticsearch';
+import { savedObjectsClientMock } from 'src/core/server/mocks';
+
+import { AlertServices } from '../../../../../alerting/server/types';
+import { Logger } from '../../../../../../../../src/core/server';
+import {
+  buildBulkBody,
+  singleBulkIndex,
+  singleSearchAfter,
+  searchAfterAndBulkIndex,
+} from './utils';
+import { SignalHit } from '../../types';
+import {
+  sampleDocNoSortId,
+  sampleSignalAlertParams,
+  sampleDocSearchResultsNoSortId,
+  sampleDocSearchResultsNoSortIdNoHits,
+  sampleDocSearchResultsWithSortId,
+  sampleEmptyDocSearchResults,
+  repeatedSearchResultsWithSortId,
+} from './__mocks__/es_results';
+import { SignalAlertParams } from './types';
+import { buildEventsSearchQuery } from './build_events_query';
+
+const mockLogger: Logger = {
+  info: jest.fn((logString: string) => {
+    return logString;
+  }),
+  warn: jest.fn((logString: string) => {
+    return logString;
+  }),
+  trace: jest.fn(),
+  debug: jest.fn((logString: string) => {
+    return logString;
+  }),
+  error: jest.fn((logString: string) => {
+    return logString;
+  }),
+  fatal: jest.fn(),
+  log: jest.fn(),
+};
+describe('utils', () => {
+  beforeEach(() => {
+    jest.clearAllMocks();
+  });
+  describe('buildBulkBody', () => {
+    test('if bulk body builds well-defined body', () => {
+      const sampleParams: SignalAlertParams = sampleSignalAlertParams(undefined);
+      const fakeSignalSourceHit = buildBulkBody(sampleDocNoSortId, sampleParams);
+      expect(fakeSignalSourceHit).toEqual({
+        someKey: 'someValue',
+        '@timestamp': 'someTimeStamp',
+        signal: {
+          '@timestamp': fakeSignalSourceHit.signal['@timestamp'], // timestamp generated in the body
+          rule_revision: 1,
+          rule_id: sampleParams.id,
+          rule_type: sampleParams.type,
+          parent: {
+            id: sampleDocNoSortId._id,
+            type: 'event',
+            index: sampleDocNoSortId._index,
+            depth: 1,
+          },
+          name: sampleParams.name,
+          severity: sampleParams.severity,
+          description: sampleParams.description,
+          original_time: sampleDocNoSortId._source['@timestamp'],
+          index_patterns: sampleParams.index,
+          references: sampleParams.references,
+        },
+      });
+    });
+  });
+  describe('singleBulkIndex', () => {
+    test('create successful bulk index', async () => {
+      const sampleParams = sampleSignalAlertParams(undefined);
+      const sampleSearchResult = sampleDocSearchResultsNoSortId;
+      const savedObjectsClient = savedObjectsClientMock.create();
+      const bulkBody = sampleDocSearchResultsNoSortId.hits.hits.flatMap(doc => [
+        {
+          index: {
+            _index: process.env.SIGNALS_INDEX || '.siem-signals-10-01-2019',
```

Comment:
> `'.siem-signals-10-01-2019'`

Since we are going to be changing this with an upcoming PR as we finalize the index, how about we make this a const right in here:

```ts
x-pack/legacy/plugins/siem/common/constants.ts
```

Something in that file like this:

```ts
const DEFAULT_SIGNALS_INDEX = .siem-signals;
```

We should just drop the date time of `-10-01-2019` in the name and then update everywhere we see this string to use that constant? It would make it easier to find all the spots and change them out when we change out the signals index way.

Looking closer though... 

An additional thing we could do (as an aside) ... Just in case if we get the requirement to make the signals index output user driven/configurable ... Would be to make our functions which use the signals index "pure" and move our "impure" reading of the environment variable up more closer to the icky side effecting part of the "handler" for routes.

That's also better for your tests as you can inject what you want, and technically functions which read something global such as `env` are not pure ... 

So, for example, 

```ts
export const singleBulkIndex = async (
  sr: SignalSearchResponse,
  params: SignalAlertParams,
  service: AlertServices,
  logger: Logger
): Promise<boolean> => {
```

Would just become something like:

```ts
export const singleBulkIndex = async (
  sr: SignalSearchResponse,
  params: SignalAlertParams,
  service: AlertServices,
  index: string,
  logger: Logger
): Promise<boolean> => {
```

And then all calling code above this will keep pushing down index until you get right to the handler. If you feel the arguments are too long are hard to read, then you can make them into an options object like we have in other spots.

This would mirror the `buildEventsReIndex({` API where it injects the index into its self. Lifting state and side effects up more towards the top of the application typically leads to better testable pure functions and an overall better and easier time to adapt code to changing requirements. Also the access of the environment would be grouped together in just one or two files (hopefully).  

Let me know what you think. This part is optional and can be a follow on PR, but seeing these tests, this might be a great time to just do it and get it out of the way.